[Gnoh Cheng Yi] iP

GnohChengYi opened the pull request:
# BluePro

> “Don't comment bad code ― rewrite it” 
> — Brian W. Kernighan and P. J. Plaugher ([source](https://en.wikipedia.org/wiki/The_Elements_of_Programming_Style))

BluePro aims to help you simplify stuff. It is

* text-based
* easy to learn
* _~~SUPER~~_ FAST to use

What are you waiting for? Do the following now!

1. Download BluePro from [here](https://github.com/GnohChengYi/ip/releases)
2. Open a command window in the folder where the jar file is
3. Run the command `java -jar ip.jar`

And it is **_FREE_**!

Features:

* [x] Manage tasks
* [x] 3 types of tasks: ToDo, Deadline, Event
* [ ] Reminders (coming soon)

If you know Java, feel free to fork and create your unique BluePro! Here's the `main` method:

```java
public class Blue {
    public static void main(String[] args) {
        new Blue("data/tasks.txt").run();
    }
}
```

Happy coding! 😁

Commit:
Add Gradle support

GnohChengYi committed:
Rename the name of chatbot to Blue, and update files that contain Duke

GnohChengYi committed:
Greet the user, echo commands entered by the user, and exits when the user types "bye"

GnohChengYi committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested

GnohChengYi committed:
Add the ability to mark tasks as done

GnohChengYi committed:
Add support for tracking ToDos, Deadlines, and Events

GnohChengYi committed:
Redirect input/output to semi-automate the testing of Blue

GnohChengYi committed:
Teach Blue to deal with errors

GnohChengYi committed:
Add support for deleting tasks from the list

GnohChengYi committed:
Abstract out command handling for ToDos, clean up code

GnohChengYi committed:
Abstract out more command handlings and create BlueException

GnohChengYi committed:
Complete the abstraction of command handlers

GnohChengYi committed:
Save tasks in hard disk whenever task list changes

GnohChengYi committed:
Change the format of save file

In the previous format, there is no good way to definitively construct a
Task, the user could have used brackets and spaces etc. in the title.

The new format is:
* First line represents the type of Task
* Second line represents the title
* Third line, if there is any, represents the by (for Deadline) or at (for
Event)
* Use empty line to separate between Tasks

Using TaskMapper to handle the conversion between Task and its
String representation makes the code cleaner.

GnohChengYi committed:
Use an independent function to handle the creation of files

GnohChengYi committed:
Teach Blue how to understand dates and times

GnohChengYi committed:
Merge branch 'branch-Level-7'

GnohChengYi committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java

GnohChengYi committed:
Fix compile errors cause by LocalDateTime after merging

GnohChengYi committed:
Extract a classes Ui and Storage

- Ui: deals with interactions with the user
- Storage: deals with loading tasks from the file and saving tasks in the file

GnohChengYi committed:
Extract Parser to deal with making sense of user commands

GnohChengYi committed:
Divide classes into packages blue, blue.task and blue.handler

GnohChengYi committed:
Update README.md

Update the location of Blue.java

GnohChengYi committed:
Add JUnit tests to test Parser and TaskList

GnohChengYi committed:
Package the app as an executable JAR file

GnohChengYi committed:
Merge branch 'master' of https://github.com/GnohChengYi/ip

GnohChengYi committed:
Add JavaDoc comments to the code

GnohChengYi committed:
Tweak the code to comply with the coding standard

GnohChengYi committed:
Give users a way to find a task by searching for a keyword

GnohChengYi committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/blue/TaskList.java

GnohChengYi committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/blue/Blue.java

GnohChengYi committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

GnohChengYi committed:
Add checkstyle files

GnohChengYi committed:
Fix checkstyle violations to prepare for build

GnohChengYi committed:
Update blue.gradle so that the built jar works properly

GnohChengYi committed:
Merge branch 'branch-A-Gradle'

GnohChengYi committed:
Set up Java FX

GnohChengYi committed:
Write simple "Hello World" program using Java FX

GnohChengYi committed:
Create basic GUI for Blue

GnohChengYi committed:
Abstract out graphic interactions with user, use Blue to launch

GnohChengYi committed:
Display correct response text in GUi

GnohChengYi committed:
Use FXML, based largely on JavaFX tutorial part 4 of SE-EDU

Tutorial: https://se-education.org/guides/tutorials/javaFxPart4.html

GnohChengYi committed:
Merge branch 'branch-Level-10'

xyliew25 commented on src/main/java/blue/Blue.java:
```diff
@@ -0,0 +1,138 @@
+public class Blue {
+    private static final String CONFUSED_RESPONSE =
+            "☹ OOPS!!! I'm sorry, but I don't know what that means :-(";
+    private final Storage storage;
```
Constants should be in caps.

GnohChengYi replied:
I think constants refer to only `static final`. This is not specifically mentioned in the coding style guide of SE-EDU, but it asks us to follow the [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names) for  anything not mentioned. 

In the guide, it states that:
> Constants are static final fields whose contents are deeply immutable and whose methods have no detectable side effects. This includes primitives, Strings, immutable types, and immutable collections of immutable types. If any of the instance's observable state can change, it is not a constant. Merely intending to never mutate the object is not enough.

So, I think the naming of `storage` is fine here.

xyliew25 commented on src/main/java/blue/Blue.java:
```diff
@@ -0,0 +1,138 @@
+            "☹ OOPS!!! I'm sorry, but I don't know what that means :-(";
+    private final Storage storage;
+    private TaskList tasks;
+    private final Ui ui;
```
Same here, constants should be in caps. Generally, variable names with the `final` keyword should be in caps. I saw there is a few more such violation in the other files.

xyliew25 commented on src/main/java/blue/Blue.java:
```diff
@@ -0,0 +1,138 @@
+        FindHandler findHandler = new FindHandler(tasks);
+        ExitHandler exitHandler = new ExitHandler(tasks);
+
+        // put the handlers into HashMap
```
Not sure if this is a coding standard violation, but I think the first letter of the first word of a comment can be in caps to ensure consistency. This part is probably overlooked.

GnohChengYi replied:
I can't find this mentioned anywhere in SE-EDU coding standard and Google Java Style Guide, so capitalizing the first letter in a comment is probably not a coding standard. (If you can find it, let me know).

But I think you have a valid point, that capitalizing the first letter makes it more consistent.

xyliew25 commented on src/main/java/blue/Parser.java:
```diff
@@ -0,0 +1,32 @@
+ * Helps in making sense of the user input.
+ */
+public class Parser {
+    static String getCommand(String input) {
```
Not sure if this is a coding standard violation as well and not sure if you intended to make this method package-private, else I think you can add in the appropriate access modifier. If it is the former, I saw there are a few such cases in the other files as well.

GnohChengYi replied:
Yup, I intend to make everything as private as possible, that's why there are many methods with only package-level access.

xyliew25 commented on src/main/java/blue/Blue.java:
```diff
@@ -0,0 +1,138 @@
+        commandHandlers.put(Command.EXIT, exitHandler);
+    }
+
+    private boolean canHandle(String input) {
```
This method name sounds, to me, like a boolean rather than a verb. Perhaps a suggestion eg. `checkCanHandle`

xyliew25 replied:
Regarding this, I reviewed and I think it's alright to name a method returning boolean to sound like a boolean, as per the sample code given under MoreOOP where there is a `isExit()` method in the `Command` class.

GnohChengYi replied:
I am actually not quite sure about how to name this method. `canHandle` sounds like only checking whether it can handle, but it is actually also handling it, which makes the name misleading.

The `boolean` was needed to determine whether it is time to exit the application. I am not sure whether it is better to handle the exit command at the GUi level or at the level of the bot logic. I might improve the design if I found a better way.

xyliew25 reviewed (commented):
Overall, it was a job well done! Just a few minor issues. Do correct me if I misunderstood or commented something wrongly :)

xyliew25 commented on src/main/java/blue/Command.java:
```diff
@@ -0,0 +1,15 @@
+    public static final String DELETE = "delete";
+    public static final String EXIT = "bye";
+    public static final String FIND = "find";
+}
```
Perhaps can consider using enums?

xyliew25 commented on src/main/java/blue/task/Task.java:
```diff
@@ -0,0 +1,49 @@
+        }
+        string += "] " + title;
+        return string;
+    }
```
```suggestion
    public String toString() {
        String taskStatus = isDone ? "X" : " ";
        return "[" + taskStatus + "] " + title;
    }
```

xyliew25 commented on src/main/java/blue/task/ToDo.java:
```diff
@@ -0,0 +1,29 @@
+     */
+    public static String getClassRepr() {
+        return "T";
+    }
```
Perhaps can replace with a public static constant property eg. `public static final CLASS_REPR = "T";`, as there is no function logic present? Similar for Event and Deadline class.

xyliew25 commented on src/main/java/blue/task/Task.java:
```diff
@@ -0,0 +1,49 @@
+        this(title, false);
+    }
+
+    /**
+     * Constructs a Task.
+     *
+     * @param title  Title of task.
+     * @param isDone Whether the task is done.
+     */
+    public Task(String title, boolean isDone) {
+        this.title = title;
+        this.isDone = isDone;
+    }
```
Hm may I know what is this overloaded constructor for? As I noticed this overloaded constructor is not called anywhere, except the other constructor. Also, personally, I don't think users will add a task that is already done.

xyliew25 commented on src/main/java/blue/handler/FindHandler.java:
```diff
@@ -0,0 +1,36 @@
+        List<Task> foundTasks = new ArrayList<>();
+        String keyword = Parser.getArguments(input)[0];
+        for (Task task : taskList.getAll()) {
+            if (task.getTitle().contains(keyword)) {
```
This is a functionality issue rather than a code quality issue. It seems that the keyword is only searched on the task title. Perhaps you can just search the whole string so as to allow users to search for dates as well ie. `at` from `Event` and `by` from `Deadline`.

xyliew25 commented on src/main/java/blue/handler/ExitHandler.java:
```diff
@@ -0,0 +1,16 @@
+package blue.handler;
+
+import blue.TaskList;
+
+public class ExitHandler extends CommandHandler {
+    private static final String EXIT_CONTENT = "Bye! (Yay! Time to rest!)";
+
+    public ExitHandler(TaskList taskList) {
+        super(taskList);
+    }
+
+    @Override
+    public String handle(String input) {
+        return EXIT_CONTENT;
+    }
+}
```
If I'm not wrong, I think this class will never be used as if the `EXIT` command is being detected in the `canHandle()` method in `Blue` class, `ui` will `sayGoodbye()` and `shouldContinue` is `false` thus breaking out of the while loop.

xyliew25 reviewed (commented):
Very sorry! I thought I was supposed to review the coding standards on your pull request. I've added the comments for coding quality as well.

GnohChengYi committed:
Update src/main/java/blue/task/Task.java

Shorten if-else statements to a single line, as suggested by @xyliew25

Co-authored-by: Liew Xin Yi <75534016+xyliew25@users.noreply.github.com>

GnohChengYi committed:
Use assert feature to document important assumptions

GnohChengYi committed:
Blue class: improve code quality

Blue class has:
* deeply nested method
* magic literal
* dead code
* unnecessary class attributes
which negatively impacts the code quality.

Let's refactor them.

GnohChengYi committed:
Command class: refactor into Command enum

Command class contains String constants for the commands. This reduces the flexibility of the
chatbot to specify which inputs match the commands.

Let's change the Command class into a Command enum, and pass the responsibility of matching the
user inputs to commands to the Parser class.

Letting the Parser class to handle the matching of inputs to commands is appropriate because it
has a getCommand method which is called by the Blue class to get the Command. The level of
abstraction is also consistent within the Parser class.

GnohChengYi committed:
Improve code quality

GnohChengYi committed:
Improve code quality and fix saving bug

Code quality:
* Delete dead code
* Use constants to replace methods for task representations - suggested by @xyliew25
* Replace String literals with constants

Save is not called anywhere, so the tasks are not saved every time after modifying task list.
Let's let the Blue class to be the intermediary between GUI and Storage.

Text-based UI is no longer needed. Let's remove the Ui class.

GnohChengYi committed:
Use the Stream feature of Java to handle finding tasks

GnohChengYi committed:
Merge pull request #1 from GnohChengYi/branch-A-Assertions

Use assert feature to document important assumptions

GnohChengYi committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/blue/Blue.java
#	src/main/java/blue/MainWindow.java
#	src/main/java/blue/Parser.java
#	src/main/java/blue/Storage.java

GnohChengYi committed:
Merge branch 'branch-A-Streams'

# Conflicts:
#	src/main/java/blue/handler/FindHandler.java

GnohChengYi committed:
Show the total number of tasks for each type of task

GnohChengYi committed:
Merge branch 'branch-C-Statistics'

GnohChengYi committed:
Give credit to AB3 code (JavaFX tutorial)

GnohChengYi committed:
Add representative screenshot and fix bug of not saving done status

GnohChengYi committed:
Set theme jekyll-theme-architect

GnohChengYi committed:
Update method header comment to follow the coding standard

GnohChengYi committed:
Merge branch 'master' of https://github.com/GnohChengYi/ip

Choose theme for GitHub page.

GnohChengYi committed:
Add a user guide

GnohChengYi committed:
Fix bug of unable to delete last task, and improve code

GnohChengYi committed:
Merge branch 'master' of https://github.com/GnohChengYi/ip

GnohChengYi committed:
Improve code quality - remove deeply nested code
